### PR TITLE
[WIP] New API endpoint: UpdateVlanIpRange

### PR DESCRIPTION
**Pull request by kioie**
## Description

Following #3111
A new endpoint that will allow the shrinking and extending of public IP range

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## Screenshots (if appropriate):
Screen Shot 2019-07-23 at 6 55 38 PM

Screen Shot 2019-07-23 at 6 56 33 PM

Screen Shot 2019-07-23 at 6 57 19 PM

Screen Shot 2019-07-23 at 6 57 57 PM

## How Has This Been Tested?

Run tests for shrinking IP range to not include IPs that have already been used.
Run tests for invalid IPs(invalid format, invalid length etc)
Run tests for gateway conflicts

### CODE REVIEW

**Comment by kioie** [comment]
@DaanHoogland @wido @svenvogel opened a new pull request as previous one had merge issues

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 1]
```diff
@@ -3958,6 +3960,116 @@ public VlanVO doInTransaction(final TransactionStatus status) {
...
+
+        if(endIp == null){
+            endIp= currentEndIP;
+        }
+
+        // Check if the VLAN has any allocated public IPs
+        final List<IPAddressVO> ips = _publicIpAddressDao.listByVlanId(id);
+        for (final IPAddressVO ip : ips) {
+            if (ip.getState()== IpAddress.State.Allocated){
+                if (!Strings.isNullOrEmpty(startIp) && NetUtils.ip2Long(startIp) > NetUtils.ip2Long(ip.getAddress().addr())) {
```
Comment (kioie) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java]: @DaanHoogland did you suggest that it should report all IPs that could potentially be a problem in one go? Both upper and lower that is? I'm I right?

Comment (DaanHoogland) [reply server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java]: Hey @kioie I answerred in the old PR but let's continue here. I think mentioning the upper and lower and maybe the number of addresses should give an admin enough info. I am not sure if we should block for that or do further work in a new PR. I'll leave that to you .

Comment (kioie) [reply server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java]: Extended vlan checks, also slightly restructed code

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 2]
```diff
@@ -1727,6 +1730,38 @@ private void checkZoneParameters(final String zoneName, final String dns1, final
         }
     }
 
+    private void checkAllocatedIpsAreWithinVlanRange(List<IPAddressVO> ips, String startIp, String endIp){
```
Comment (kioie) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java]: Added this method

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 3]
```diff
@@ -1727,6 +1730,38 @@ private void checkZoneParameters(final String zoneName, final String dns1, final
...
+            if (!Strings.isNullOrEmpty(endIp) && NetUtils.ip2Long(endIp) < NetUtils.ip2Long(allocatedIP.getAddress().addr())) {
+                throw new InvalidParameterValueException("The start IP address must have a lower IP address value " +
+                        "than "+ listAllocatedIPs.get(0).getAddress() + " which is already in use. The end IP must have a " +
+                        "higher IP address than "+  allocatedIP.getAddress()+" .IPs " +
+                        "already allocated in this range: "+listAllocatedIPs.size());
+            }
+        }
+    }
+
+    private void checkGatewayOverlap(String startIp, String endIp, String gateway){
```
Comment (kioie) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java]: and this one

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 4]
```diff
@@ -3958,6 +3993,98 @@ public VlanVO doInTransaction(final TransactionStatus status) {
...
+            startIp= currentStartIP;
+        }
+
+        if(endIp == null){
+            endIp= currentEndIP;
+        }
+
+        // Check if the VLAN has any allocated public IPs
+        final List<IPAddressVO> ips = _publicIpAddressDao.listByVlanId(id);
+        checkAllocatedIpsAreWithinVlanRange(ips,startIp,endIp);
```
Comment (kioie) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java]: updated

File: api/src/main/java/com/cloud/configuration/ConfigurationService.java [hunk 5]
```diff
@@ -269,8 +270,25 @@
...
-    Vlan createVlanAndPublicIpRange(CreateVlanIpRangeCmd cmd) throws InsufficientCapacityException, ConcurrentOperationException, ResourceUnavailableException,
-        ResourceAllocationException;
+    Vlan createVlanAndPublicIpRange(CreateVlanIpRangeCmd cmd) throws InsufficientCapacityException,
+            ConcurrentOperationException, ResourceUnavailableException, ResourceAllocationException;
+
+    /**
+     * Updates the IP address Range for the VLAN on the database, a
+     *
+     * @param cmd
+     * @param vlanId
```
Comment (DaanHoogland) [inline api/src/main/java/com/cloud/configuration/ConfigurationService.java]: this javadoc is not actual. only @param cmd is present.

File: api/src/main/java/com/cloud/configuration/ConfigurationService.java [hunk 6]
```diff
@@ -269,8 +270,25 @@
...
+     * @param gateway
+     * @param startIP
+     * @param endIP
+     * @param netmask
+     * @throws com.cloud.exception.ConcurrentOperationException
```
Comment (DaanHoogland) [inline api/src/main/java/com/cloud/configuration/ConfigurationService.java]: please add comment for reason..?

**Comment by rhtyd** [comment]
Manual testing needed.

**Comment by andrijapanicsb** [comment]
@blueorangutan package

**Comment by andrijapanicsb** [comment]
@blueorangutan test

**Comment by andrijapanicsb** [comment]
Trillian tests cancelled.

@kioie I've tested manually and confirmed IP addresses are NOT removed from the user_ip_address table, that I recall me/someone has asked about.

Kindly fix by implementing row removals from this table when the IPs are "removed" from the range.

Also what about updating ipv6 ranges? kindly ping @wido when/if you add this

**Comment by rhtyd** [comment]
Let's revisit this after 4.14 @andrijapanicsb

**Comment by kioie** [comment]
@DaanHoogland @andrijapanicsb @svenvogel @borisstoyanov Hey guys, cleaning this up on a new PR(3997), apologies for the mess.

**Pull request closed without merging**